### PR TITLE
Refactor DAO using OOP, add candidates API and fix a lot of DB and API code

### PR DESCRIPTION
**Pull request by LeanidShutau**

### CODE REVIEW

File: server/routes/hirings.route.js [hunk 1]
```diff
@@ -1,11 +1,13 @@
 const controller = require('./../controllers/hirings.controller');
 
-const hirings = (app) => {
+function init(app) {
   app.post('/hirings', controller.writeHiring);
```
Comment (4en1x) [inline server/routes/hirings.route.js]: controller.createHiring

File: server/controllers/hirings.controller.js [hunk 2]
```diff
@@ -1,115 +1,96 @@
...
   } catch (err) {
     return res.status(500).end();
   }
 }
 
 async function deleteHiring(req, res) {
   try {
-    await hiringsDB.deleteHiring(req.params.id);
-    return res.status(200).end();
+    await db.hirings.deleteHiring(req.params.id);
```
Comment (4en1x) [inline server/controllers/hirings.controller.js]: db.hirings.delete

File: server/config.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+const config = {};
```
Comment (aaneitchik) [inline server/config.js]: I think this config would be easier to read if it was just
```sh
const config = {.....};
module.exports = config;
```

File: server/controllers/candidates.controller.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    const candidates = await db.candidates.readPage(req.params.p);
+    res.json(candidates);
+  } catch (err) {
+    res.status(500).end();
+  }
+}
+
+async function readCandidate(req, res) {
+  try {
+    const candidate = await db.candidates.readOne(req.params.id);
```
Comment (aaneitchik) [inline server/controllers/candidates.controller.js]: Don't you need to do something if candidate was found successfully?

File: server/controllers/interviews.controller.js [hunk 5]
```diff
@@ -1,69 +1,61 @@
...
     return res.status(500).end();
   }
 }
 
 async function readInterviews(req, res) {
   const actions = {
-    my: interviewsDB.getMyInterviews,
-    assigned: interviewsDB.getAssignedInterviews,
-    all: interviewsDB.getAllInterviews,
+    my: db.interviews.readPageToUser,
```
Comment (aaneitchik) [inline server/controllers/interviews.controller.js]: readPageToUser, readPageFromUser - wolud be cool if the names were more clear, I can't guess what they do

File: server/dao/feedbacks.dao.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      sql: `SELECT * FROM ${this.table}
+            WHERE interview_id = ? AND user_id = ?`,
+      values: [interviewId, userId],
+    });
+
+    feedback.fields = await readFields(feedback.id);
+    return feedback;
+  }
+
+  async readAllFromInterview(id) { // TEST IT!
```
Comment (aaneitchik) [inline server/dao/feedbacks.dao.js]: There is some sort of convention for comments like this
```sh
// TODO: TEST IT!
```
Some IDEs (Webstorm, for example) recognize them, and these comments are easier to track in the code

File: server/dao/feedbacks.dao.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    const feedbacks = await Promise.all(this.connection.queryAsync({
+      sql: `SELECT * FROM ${this.table} WHERE interview_id = ?`,
+      values: [id],
+    }).map(async (feedback) => {
+      feedback.fields = await readFields(feedback.id);
```
Comment (aaneitchik) [inline server/dao/feedbacks.dao.js]: Of course, this is not critical, but it is considered bad practice to use .map (not only .map, but some other array methods as well) as mutable. If you're mutating the original object, use something like forEach instead.

File: server/dao/interviews.dao.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+      addition: joins + where,
+      limit,
+      values: [id, (page - 1) * this.top, this.top],
+    });
+
+    return interviews;
+  }
+
+  async readPageFromUser(id, page = 1) {
+    const fields = 'interviews.id", type, date, place';
```
Comment (aaneitchik) [inline server/dao/interviews.dao.js]: This double quote here probably is a typo?

File: server/services/candidates.service.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
+const fecha = require('fecha');
+
+function buildCandidate(reqBody) {
+  const candidate = {};
+  let links = [];
+  let city;
+  Object.keys(reqBody).forEach((key) => {
```
Comment (aaneitchik) [inline server/services/candidates.service.js]: Building an object from some array seems like a job for .reduce() :)

File: server/utils.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
+function amplifyParams(src, def) {
+  const result = {};
+  Object.keys(def).forEach((param) => {
+    if (!Object.prototype.hasOwnProperty.call(src, param)) {
```
Comment (aaneitchik) [inline server/utils.js]: 1. You shouldn't write it with call and prototype, it shoud be just src.hasOwnProperty(param)
2. Better write 'default', do not use shortcuts like 'def'.
2. If I'm not mistaken, you can simply replace this function body with
```sh
return Object.assign({}, def, src);
```
Empty object is needed to not mutate def object.

Comment (LeanidShutau) [inline server/utils.js]: 1.  http://eslint.org/docs/rules/no-prototype-builtins
2. "default" is a keyword in JS for switch statement

File: server/controllers/interviews.controller.js (hunk 5 above)
Comment (ala-n) [inline server/controllers/interviews.controller.js]: Absolutely agree
PS: that work for future refactoring, not for that PR.

File: server/services/hirings.service.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+function createHiringObject(req) {
+  return toSnake({
+    userId: req.user.id,
+    dateOpen: fecha.format(new Date(), 'YYYY-MM-DD HH:mm:ss'),
+    candidateId: req.query.candidate,
+  });
+}
+
+function createHiringUpdateObject(reqBody) {
+  if (reqBody.vacancyId) {
```
Comment (4en1x) [inline server/services/hirings.service.js]: I dont think that we should check only this argument, what about candidate_id?

File: server/utils.js (hunk 10 above)
Comment (ala-n) [inline server/utils.js]: +100500 thousands for third point.

File: server/controllers/interviews.controller.js (hunk 5 above)
Comment (LeanidShutau) [inline server/controllers/interviews.controller.js]: As I can't guess, what getMy... and getAssigned... do :)
readPageToUser means: "read one page of interviews assigned to user".
readPageFromUser means: "read one page of interviews created by user".
I have no idea how to name those methods better, butWithoutTooLongNamesThatCanNotFitInMyScreen

File: server/dao/basic.dao.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+      values: [resource, id],
+    });
+  }
+
+  async delete(id) {
+    const data = await connection.queryAsync({
+      sql: `DELETE FROM ${this.table} WHERE id = ?`,
+      values: [id],
+    });
+
```
Comment (4en1x) [inline server/dao/basic.dao.js]: Maybe something like 
if (data.affectedRows === 0) {
      throw new Error('Already deleted');
}

File: server/services/hirings.service.js (hunk 11 above)
Comment (LeanidShutau) [inline server/services/hirings.service.js]: There is no mistake here. All is OK

File: server/dao/interviews.dao.js [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+            sql: 'INSERT INTO feedback_fields SET ?',
+            values: [field],
+          });
+        }));
+      }));
+
+      await this.connection.commit();
+      return id;
+    } catch (err) {
+      return this.connection.rollback(() => {
```
Comment (4en1x) [inline server/dao/interviews.dao.js]: I think it must be like this
throw new Error()
return this.connection.rollback()

For now app crashes  on MySQL error

Comment (LeanidShutau) [inline server/dao/interviews.dao.js]: Throwing an error terminates code exec., so you can not rollback after it.
I sure,
`catch (err) {
  await this.connection.rollbackAsync();
  throw err;
}`
is a solution

Comment (4en1x) [inline server/dao/interviews.dao.js]: No, it not) app still crash

File: server/dao/candidates.dao.js [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+        throw err;
+      });
+    }
+  }
+
+  async readPage(page = 1) {
+    try {
+      await this.connection.beginTransactionAsync();
+
+      const fields = `${this.table}.id, name, surname, primary_skill, status, last_change_date, city.name AS city`;
```
Comment (4en1x) [inline server/dao/candidates.dao.js]: That's the right solution:
const fields = `${this.table}.id, ${this.table}.name, ${this.table}.surname, ${this.table}.primary_skill, ${this.table}.status, ${this.table}.last_change_date, cities.name AS city`;

File: server/controllers/candidates.controller.js (hunk 4 above)
Comment (4en1x) [inline server/controllers/candidates.controller.js]: res.json(candidate);

File: server/services/candidates.service.js [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    if (key === 'city') {
+      city = reqBody[key];
+      return;
+    }
+    candidate[key] = reqBody[key];
+  });
+
+  return {
+    candidate,
```
Comment (4en1x) [inline server/services/candidates.service.js]: toCamel(candidate)

Comment (4en1x) [inline server/services/candidates.service.js]: And you should add field user_id

File: server/services/candidates.service.js (hunk 9 above)
Comment (aaneitchik) [inline server/services/candidates.service.js]: The name is not really good, it's not only building a candidate. I don't see why this isn't .reduce

File: server/utils.js (hunk 10 above)
Comment (aaneitchik) [inline server/utils.js]: @LeanidShutau 
1. Ok, didn't know about that one.
2. Ok, let it be defaultOpts or something like that, but certainly not def. def would be the name of some var in C++, not JS

**Pull request merged**